README.md
### Overview
HyperDocs is an Android SDK of HyperVerge's Artificial Intelligence Driven KYC Digitisation (OCR) for PAN Cards, Aadhaar cards and Passports (Driving Licenses and Voting Cards - coming soon)

HyperDocs is deep-learning based mobile SDK for OCR that integrates into your existing mobile applications (banking, financial services, telecom, etc.) in order to capture the customer's KYC document (PAN Card, Aaadhar, and Passport) and extracts the details from them with greater than 99% accuracy.

### Prerequisites
- Gradle Version: 4.4 (Recommended)
- minSdkVersion 14
- targetSdkVersion 27

### Sample
- Please refer to the sample app provided in the repo to get an understanding of the implementation process.
- To run the app, clone/download the repo and open **sample** using latest version of Android Studio
- Open project build.gradle and replace **aws_access_key** and **aws_secret_pass** with the credentials provided by HyperVerge
- Open `Configs.java` and assign **appId** and **appKey** provided by HyperVerge to APP_ID and APP_KEY respectively
- Build and run the app

#### Integration Steps
- Add the following set of lines to your `app/build.gradle`

```groovy
android {
    defaultConfig {
        renderscriptTargetApi 19
        renderscriptSupportModeEnabled true
    }
}
dependencies {
    compile('co.hyperverge:hyperdocssdk:1.1.0@aar', {
        transitive = true
    })
}
```
- Add the following set of lines to the Project (top-level) `build.gradle`

```groovy
allprojects {
    repositories {
        maven {
            url "s3://hyperdocs-sdk/android/releases"
            credentials(AwsCredentials) {
                accessKey "aws_access_key"
                secretKey "aws_secret_pass"
            }
        }
    }
}
```
where **aws_access_key** and **aws_secret_pass** will be given by HyperVerge
- **Permissions**: The app requires the following permissions to work.
    - *Camera*
    - *Autofocus*
    - *Read & Write External Storage*
    - *Internet*
    - *Access Network State*

    Kindly note that for android v23 (Marshmallow) and above, you need to handle the runtime permissions inside your app.

- Add the following line to your Application class (the class which extends android.app.Application) for initializing our Library. This must be run only once. Check [this](https://guides.codepath.com/android/Understanding-the-Android-Application-Class) link if you are unsure of what an Application class is. Here appId and appKey are the credentials that have been supplied by us.
```java
HyperDocsSDK.init(context, appId, appKey);
```

- For clicking photo of Id Card(Passport, Pan or Aadhaar) and reading text on it using our sophisticated OCR technology, call the corresponding function based on document type from your app. This will start a camera activity where you can click front, back or front+back images of the card, which in turn will be sent to our OCR for recognition of text in different fields and you can access it by implementing the above mentioned `CameraActivity.ImageListener()` in your code.
    - **Passport**

        ```java
        CameraActivity.startForPassport(context, passportFrontImagePath, passportBackImagePath, documentSide, new CameraActivity.ImageListener() {
            @Override
            public void onOCRComplete(JSONObject result) {
                //result JSONObject will have the detected text and the other information needed. We have documented each key in the object below
                //Implement relevant logic from the info detected by our OCR here
            }
        });
        ```
    - **Aadhaar**

        ```java
        CameraActivity.startForAadhaar(context, aadhaarFrontImagePath, aadhaarBackImagePath, documentSide, new CameraActivity.ImageListener() {
            @Override
            public void onOCRComplete(JSONObject result) {
                //result JSONObject will have the detected text and the other information needed. We have documented each key in the object below
                //Implement relevant logic from the info detected by our OCR here
            }
        });
        ```
    - **Pan Card**

        ```java
        CameraActivity.startForPan(context, panImagePath, new CameraActivity.ImageListener() {
            @Override
            public void onOCRComplete(JSONObject result) {
                //result JSONObject will have the detected text and the other information needed. We have documented each key in the object below
                //Implement relevant logic from the info detected by our OCR here
            }
        });
        ```

    where
    - `documentSide` is a variable of enum type `CameraActivity.DocumentSide` and can have `CameraActivity.DocumentSide.FRONT_BACK`, `CameraActivity.DocumentSide.FRONT` and `CameraActivity.DocumentSide.BACK` as possible values
    - `passportFrontImagePath`, `passportBackImagePath`, `aadhaarFrontImagePath`, `aadhaarBackImagePath` & `panImagePath` are the path to the image where we will be storing the image clicked by user. For example, if `passportFrontImagePath` parameter in startForPassport method is set to `/storage/emulated/0/hyperdocs/1498028857114_front.jpg`, then the front image of the passport will be saved to this path.

- Keys in the result JSONObject passed as a parameter to onOCRComplete method of CameraActivity.ImageListener interface arranged by the type of document:
    - **Passport**
        - **passport_num**:    Passport Number
        - **given_name**:    Name of the Owner
        - **surname**:    Surname of the Owner
        - **spouse**:    Name of spouse of the Owner
        - **dob**:    Date of Birth of the Owner
        - **gender**:    Gender of the Owner
        - **doi**:    Date of Issue of Passport
        - **doe**:    Date of expiry of the Passport
        - **country_code**:    Country Code as mentioned in the Passport
        - **nationality**:    Nationality of the Owner
        - **type**:    Type as mentioned in the Passport
        - **file_num**:    File Number as mentioned in the Passport
        - **father**:    Father of the Owner
        - **mother**:    Mother of the Owner
        - **address**:    Address of the Owner as mentioned in the Passport
        - **front_url**:    Local Url of front side of the Passport image clicked by the SDK
        - **back_url**:    Local Url of back side of the Passport image clicked by the SDK
    - **Aadhaar**
        - **name**:    Name of the Owner
        - **phone**:    Phone Number of the Owner
        - **gender**:    Gender of the Owner
        - **aadhaar**:    Aadhaar Number
        - **dob**:    Date of Birth of the Owner
        - **yob**:    Year of Birth of the Owner
        - **address**:    Address of the Owner
        - **pin**:    Pin Code of the Owner's Address
        - **front_url**:    Local Url of front side of the Aadhaar image clicked by the SDK
        - **back_url**:    Local Url of back side of the Aadhaar image clicked by the SDK
    - **Pan**
        - **name**:    Name of the Owner
        - **father**:    Father's Name of the Owner
        - **surname**:    Surname of the Owner
        - **pan_no**:    Pan Number
        - **front_url**:    Local Url of front(only) side of the Pan Card image clicked by the SDK

- **Data Logging**: HyperDocs APIs, by default, donot save the idcard images sent for KYC OCR. To allow the servers to persist those images, following code should be added before calling any start method.
    
    ```
    CameraActivity.toggleDataLogging(true);
    ```
    
    Please note that this is an optional feature and, by default, it is disabled. If enabled, the saved images can help us debug the failure scenerios better and the saved images will also be feeded back to the Deep Learning OCR system so as to increase the accuracy of the system.

- **Customizations**: Some text fields, element colors, font styles, and button icons are customizable so that the look and feel of the components inside SDK can be altered to match the look and field of the app using the SDK. `Kindly note that this step is optional`. Below is the list of items that are customizable grouped by the resource file/type where the customized value/file(s) should be placed in.
    - **strings.xml**:

        ```xml
            <string name="sdk_title_text">HYPERDOCS Override</string>
            <string name="instructions_text_passport">Place Passport page inside the box Override</string>
            <string name="instructions_text_pan">Place Pan Card inside the box Override</string>
            <string name="instructions_text_aadhaar">Place Aadhaar Card page inside the box Override</string>
            <string name="preview_header_text_front_pan">FRONT PAGE REVIEW PAN Override</string>
            <string name="preview_header_text_front_passport">FRONT PAGE REVIEW PASSPORT Override</string>
            <string name="preview_header_text_back_passport">BACK PAGE REVIEW PASSPORT Override</string>
            <string name="preview_header_text_front_aadhaar">FRONT PAGE REVIEW AADHAAR Override</string>
            <string name="preview_header_text_back_aadhaar">BACK PAGE REVIEW AADHAAR Override</string>
            <string name="step_front_text_pan">Front Page of PAN Override</string>
            <string name="step_front_text_passport">Front Page of Passport Override</string>
            <string name="step_back_text_passport">Back Page of Passport Override</string>
            <string name="step_front_text_aadhaar">Front Page of Aadhaar Override</string>
            <string name="step_back_text_aadhaar">Back Page of Aadhaar Override</string>
        ```

    - **colors.xml**:

        ```xml
            <color name="camera_button_color">#273646</color>
            <color name="top_bar_color">#FFFFFF</color>
            <color name="sdk_title_text_color">#9b9b9b</color>
        ```

    - **dimens.xml**:

        ```xml
            <dimen name="preview_title_text_font_size">16sp</dimen>
            <dimen name="camera_title_text_font_size">12sp</dimen>
        ```

    - **Drawable PNGs**:
        - **ic_camera_torch_off**: Max dimension size should be less than 35dp
        - **ic_camera_torch_on**: Max dimension size should be less than 35dp
        - **ic_camera_cross**: Max dimension size should be less than 35dp

    - **styles.xml**:
        - **TextStyleHeading**: Style value for the heading text. Can be used to set font style by using it in following way:
            
            ```xml    
                <style name="TextStyleHeading">
                    <item name="fontPath">fonts/Roboto_regular.ttf</item>
                </style>
            ```
            where fontPath is the path to font `ttf` or `otf` file relative to assets folder in the app.

### Sample Code
The sample code is present in the **sample** directory of this repo. In order to use the sample app, kindly add the `appId` and `appKey` given by HyperVerge to `APP_ID` and `APP_KEY` respectively in `sample/app/src/main/java/co/hyperverge/hyperdocscombinedapp/Utils/Configs.java`.

### Contact Us
If you are interested in integrating this SDK, please do send us a mail at [contact@hyperverge.co](mailto:contact@hyperverge.co) explaining your use case. We will give you the `aws_access_key`, `aws_secret_pass`, `appId` & `appKey`, so that you can try it out.
Meanwhile, the demo app can be downloaded from Play Store by clicking [here](https://play.google.com/store/apps/details?id=co.hyperverge.hyperdocscombinedapp).


CHANGELOG.md
#### Version 1.1.0
##### - Improvements:
- Using Version 1.1 of the [KYC API](https://github.com/hyperverge/kyc-rest-api)
- Crash resolution using confidence value
- Returning of the confidence value for each extracted field
- New and better UI for capture screen
- Changed CompiledSDKVersion from 25 to 27
- Changed BuildToolsVersion from 26.0.2 to 27.0.3
- Changed TargetSDKVersion from 25 to 27


sample/app/src/test/java/co/hyperverge/hyperdocscombinedapp/ExampleUnitTest.java
package co.hyperverge.hyperdocscombinedapp;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Example local unit test, which will execute on the development machine (host).
 *
 * @see <a href="http://d.android.com/tools/testing">Testing documentation</a>
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

sample/app/src/main/java/co/hyperverge/hyperdocscombinedapp/Utils/SpUtils.java
package co.hyperverge.hyperdocscombinedapp.Utils;

import android.content.Context;
import android.content.SharedPreferences;

import org.json.JSONArray;
import org.json.JSONException;

import java.util.ArrayList;
import java.util.List;

import co.hyperverge.hyperdocscombinedapp.HyperDocsCombinedApp;

/**
 * Created by adarsh on 18/03/17.
 */
public class SpUtils {
    public static final String KEY_ORIGINAL = "keyOriginal";
    public static final String KEY_COMPRESSED = "keyCompressed";
    public static final String KEY_LIST = "keyList1";

    private static SharedPreferences sp = null;
    public static final String spInstant = "InstantSP";
    private static SharedPreferences.Editor editor = null;
    private static Boolean profileLikeSyncVariable = true;
    private static Boolean eventReceivedStreamsSyncVariable = true;

    private static SharedPreferences.Editor getEditor() {
        if (editor == null) {
            editor = getSP().edit();
        }
        return editor;
    }

    public static void init(Context context) {
        if (context != null) {
            sp = context.getSharedPreferences(spInstant, Context.MODE_PRIVATE);
            editor = sp.edit();
        }

    }

    private static SharedPreferences getSP() {
        if (sp == null) {
            init(HyperDocsCombinedApp.getInstance());
        }
        return sp;
    }

    public static void saveLists(List<String> fullList) {
        JSONArray savedList = new JSONArray();

        for(int i = 0; i < fullList.size(); i++) {
            savedList.put(fullList.get(i));
        }
        getEditor().putString(KEY_LIST, savedList.toString());
        getEditor().commit();
    }

    public static List<String> getLists() {
        String savedString = getSP().getString(KEY_LIST, "[]");
        List<String> savedArray = new ArrayList<>();

        try {
            JSONArray savedList = new JSONArray(savedString);
            for(int i = 0; i < savedList.length(); i++)
            {
                savedArray.add(savedList.getString(i));
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }
        return savedArray;
    }


}

sample/app/src/main/java/co/hyperverge/hyperdocscombinedapp/HyperDocsCombinedApp.java
package co.hyperverge.hyperdocscombinedapp;

import android.app.Application;
import co.hyperverge.hyperdocscombinedapp.Utils.Configs;
import co.hyperverge.hyperdocscombinedapp.Utils.SpUtils;
import co.hyperverge.hyperdocssdk.HyperDocsSDK;
import com.crashlytics.android.Crashlytics;
import io.fabric.sdk.android.Fabric;

/**
 * Created by sanchit on 10/05/17.
 */

public class HyperDocsCombinedApp extends Application {
    private volatile static HyperDocsCombinedApp instance = null;

    public static synchronized HyperDocsCombinedApp getInstance() {
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        Fabric.with(this, new Crashlytics());

        HyperDocsSDK.init(this, Configs.APP_ID, Configs.APP_KEY);

        SpUtils.init(this);
    }
}


sample/app/src/main/java/co/hyperverge/hyperdocscombinedapp/Activities/ImagesActivity.java
package co.hyperverge.hyperdocscombinedapp.Activities;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TableLayout;
import android.widget.TableRow;
import android.widget.TextView;

import com.bumptech.glide.Glide;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.HashMap;

import co.hyperverge.hyperdocscombinedapp.R;
import co.hyperverge.hyperdocscombinedapp.Utils.Configs;
import co.hyperverge.hyperdocssdk.workflows.ocr.activities.BaseActivity;

/**
 * Created by sanchit on 13/05/17.
 */

public class ImagesActivity extends BaseActivity {

    private String mResult;
    private LandingActivity.DocumentType documentType;
    private LinearLayout llFront;
    private LinearLayout llResultFront;
    private TextView tvName;
    private TextView tvSurname;
    private TextView tvDob;
    private TextView tvPassportNo;
    private TextView tvGender;
    private TextView tvDoi;
    private TextView tvDoe;
    private TextView tvMother;
    private TextView tvFather;
    private TextView tvAddress;
    private TextView tvSpouse;

    private ImageView ivFront;
    private ImageView ivBack;

    private LayoutInflater inflater;

    public static void start(Context context, String result, LandingActivity.DocumentType documentType) {
        Intent intent = new Intent(context, ImagesActivity.class);
        intent.putExtra("result", result);
        intent.putExtra("documentType", documentType.name());
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_images);

        if(!getIntent().hasExtra("result") || !getIntent().hasExtra("documentType"))
            return;
        mResult = getIntent().getStringExtra("result");
        documentType = LandingActivity.DocumentType.valueOf(getIntent().getStringExtra("documentType"));

        inflater = LayoutInflater.from(ImagesActivity.this);
        setupViews();
    }

    private void setTextFromJSON(JSONObject json, TextView tv, String tag)
    {
        if(json.has(tag)) {
            try {
                tv.setText(json.getString(tag));
            } catch (JSONException e) {
                e.printStackTrace();
            }
        } else {
            tv.setText("-");
        }
    }

    private void setupViews() {
        findViewById(R.id.iv_back_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                finish();
            }
        });
        JSONObject resultJSON = new JSONObject();
        try {
            resultJSON = new JSONObject(mResult);
        } catch (JSONException e) {
            e.printStackTrace();
        }

        ((TextView) findViewById(R.id.headingTV)).setText(documentType.name());

        String frontKeyList[] = new String[]{};
        String backKeyList[] = new String[]{};
        HashMap<String, String> keyToNameMap = new HashMap<>();

        if(documentType == LandingActivity.DocumentType.PAN){
            frontKeyList = Configs.PAN_ORDERED_KEYS;
            keyToNameMap = Configs.PAN_JSON_FIELD_KEY_TO_NAME_MAP();
        }
        else if(documentType == LandingActivity.DocumentType.PASSPORT){
            frontKeyList = Configs.PASSPORT_ORDERED_FRONT_KEYS;
            backKeyList = Configs.PASSPORT_ORDERED_BACK_KEYS;
            keyToNameMap = Configs.PASSPORT_JSON_FIELD_KEY_TO_NAME_MAP();
        }
        else if(documentType == LandingActivity.DocumentType.AADHAAR){
            frontKeyList = Configs.AADHAAR_ORDERED_FRONT_KEYS;
            backKeyList = Configs.AADHAAR_ORDERED_BACK_KEYS;
            keyToNameMap = Configs.AADHAAR_JSON_FIELD_KEY_TO_NAME_MAP();
        }

        int fieldsRead = 0;

        for(String key: frontKeyList){
            FieldValueTableRowViewHolder holder = new FieldValueTableRowViewHolder((TableRow) inflater.inflate(R.layout.field_value_table_row, null,false));
            holder.fieldTV.setText(keyToNameMap.get(key));
            try {
                holder.valueTV.setText(resultJSON.getString(key));
                ((TableLayout)findViewById(R.id.tl_front)).addView(holder.tableRow);
                fieldsRead++;
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        ivFront = (ImageView) findViewById(R.id.iv_front);
        ivBack = (ImageView) findViewById(R.id.iv_back);
        ImageView ivOneSide = (ImageView) findViewById(R.id.iv_one_side);
        try {
            if(resultJSON.has("front_url") && resultJSON.has("back_url")) {
                findViewById(R.id.ll_col_view).setVisibility(View.VISIBLE);
                findViewById(R.id.ll_one_side).setVisibility(View.GONE);
                Glide.with(this).load(resultJSON.getString("front_url")).placeholder(R.color.grey).into(ivFront);
                Glide.with(this).load(resultJSON.getString("back_url")).placeholder(R.color.grey).into(ivBack);
                final JSONObject finalResultJSON = resultJSON;
                ivFront.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        try {
                            showImageInFullScreenView(finalResultJSON.getString("front_url"));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                });

                ivBack.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        try {
                            showImageInFullScreenView(finalResultJSON.getString("back_url"));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
            else if(resultJSON.has("front_url")){
                findViewById(R.id.ll_col_view).setVisibility(View.GONE);
                findViewById(R.id.ll_one_side).setVisibility(View.VISIBLE);
                Glide.with(this).load(resultJSON.getString("front_url")).placeholder(R.color.grey).into(ivOneSide);
                ((TextView)findViewById(R.id.tv_one_side_image_heading)).setText("FRONT PAGE");

                final JSONObject finalResultJSON1 = resultJSON;
                ivOneSide.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        try {
                            showImageInFullScreenView(finalResultJSON1.getString("front_url"));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
            else if(resultJSON.has("back_url")){
                findViewById(R.id.ll_col_view).setVisibility(View.GONE);
                findViewById(R.id.ll_one_side).setVisibility(View.VISIBLE);
                Glide.with(this).load(resultJSON.getString("back_url")).placeholder(R.color.grey).into(ivOneSide);
                ((TextView)findViewById(R.id.tv_one_side_image_heading)).setText("BACK PAGE");

                final JSONObject finalResultJSON2 = resultJSON;
                ivOneSide.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        try {
                            showImageInFullScreenView(finalResultJSON2.getString("front_url"));
                        } catch (JSONException e) {
                            e.printStackTrace();
                        }
                    }
                });
            }
        } catch (JSONException e) {
            e.printStackTrace();
        }

        for(String key: backKeyList){
            FieldValueTableRowViewHolder holder = new FieldValueTableRowViewHolder(inflater.inflate(R.layout.field_value_table_row, null,false));
            holder.fieldTV.setText(keyToNameMap.get(key));
            try {
                holder.valueTV.setText(resultJSON.getString(key));
                ((TableLayout)findViewById(R.id.tl_front)).addView(holder.tableRow);
                fieldsRead++;
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }

        if(fieldsRead == 0){
            findViewById(R.id.emptyPlaceholderTV).setVisibility(View.VISIBLE);
        }

//        setTextFromJSON(resultJSON, tvMother, "mother");
//        setTextFromJSON(resultJSON, tvFather, "father");
//        setTextFromJSON(resultJSON, tvAddress, "address");
//        setTextFromJSON(resultJSON, tvSpouse, "spouse");

    }

    public void dismissFullScreenImageView(View view) {
        findViewById(R.id.fl_full_screen_view).setVisibility(View.GONE);
    }

    public void showImageInFullScreenView(String filePath){
        ImageView fullScreenPreviewIV = (ImageView) findViewById(R.id.iv_preview);
        Glide.with(this).load(filePath).into(fullScreenPreviewIV);
        findViewById(R.id.fl_full_screen_view).setVisibility(View.VISIBLE);
    }

    private class FieldValueTableRowViewHolder {
        public TextView fieldTV;
        public TextView valueTV;
        public TableRow tableRow;

        public FieldValueTableRowViewHolder(View itemView) {
            tableRow = (TableRow) itemView;
            fieldTV = (TextView) itemView.findViewById(R.id.tv_field);
            valueTV = (TextView) itemView.findViewById(R.id.tv_value);
        }
    }

}

sample/app/src/main/java/co/hyperverge/hyperdocscombinedapp/Utils/Configs.java
package co.hyperverge.hyperdocscombinedapp.Utils;

import android.os.Environment;

import org.json.JSONObject;

import java.util.HashMap;

/**
 * Created by sanchit on 13/05/17.
 */

public class Configs {
    public static String FILE_DIR = Environment.getExternalStorageDirectory() + "/hyperdocs/original/";

    public static String APP_ID = "";           //assign the appId given by HyperVerge to APP_ID variable
    public static String APP_KEY = "";          //assign the appKey given by HyperVerge to APP_ID variable

    public static final String[] PASSPORT_ORDERED_FRONT_KEYS = {
            "given_name",
            "surname",
            "passport_num",
            "gender",
            "dob",
            "doi",
            "doe",
            "place_of_birth",
            "place_of_issue"
    };
    public static final String[] PASSPORT_ORDERED_BACK_KEYS = {
            "father",
            "mother",
            "spouse",
            "file_num",
            "address"
    };

    public static HashMap<String, String> PASSPORT_JSON_FIELD_KEY_TO_NAME_MAP(){
        HashMap<String, String> map = new HashMap<>();
        map.put("given_name", "Name");
        map.put("surname", "Surname");
        map.put("gender", "Gender");
        map.put("passport_num", "Passport Number");
        map.put("dob", "Date of Birth");
        map.put("doe", "Date of Expiry");
        map.put("doi", "Date of Issue");
        map.put("mother", "Mother");
        map.put("father", "Father");
        map.put("address", "Address");
        map.put("spouse", "Spouse");
        map.put("place_of_issue", "Place of Issue");
        map.put("place_of_birth", "Place of Birth");
        map.put("file_num", "File Number");
        return map;
    }

    public static final String[] AADHAAR_ORDERED_FRONT_KEYS = {
            "name",
            "phone",
            "aadhaar",
            "gender",
            "dob",
            "yob",
            //"address",
            //"pin"
    };
    public static final String[] AADHAAR_ORDERED_BACK_KEYS = {
            //"aadhaar",
            "address",
            "pin"
    };

    public static HashMap<String, String> AADHAAR_JSON_FIELD_KEY_TO_NAME_MAP(){
        HashMap<String, String> map = new HashMap<>();
        map.put("name", "Name");
        map.put("phone", "Phone No");
        map.put("gender", "Gender");
        map.put("aadhaar", "Aadhaar Number");
        map.put("dob", "Date of Birth");
        map.put("yob", "Year of Birth");
        map.put("address", "Address");
        map.put("pin", "Pin Code");
        return map;
    }

    public static final String[] PAN_ORDERED_KEYS = {
            "name",
            "father",
            "pan_no",
            "date"
    };

    public static HashMap<String, String> PAN_JSON_FIELD_KEY_TO_NAME_MAP(){
        HashMap<String, String> map = new HashMap<>();
        map.put("name", "Name");
        map.put("father", "Father");
        map.put("pan_no", "Pan Number");
        map.put("date", "Date of Birth");
        return map;
    }
}
